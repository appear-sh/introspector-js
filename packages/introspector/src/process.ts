import { schemaFromValue } from "./contentTypes/contentTypes.js"
import { StringSchemaType } from "./contentTypes/jsonSchema.types.js"
import { isNonNullable } from "./helpers.js"
import { Operation } from "./report.js"

const getBodySchema = async (input: Request | Response) => {
  const clone = input.clone()
  if (!clone.body) return null

  const contentMediaType = clone.headers
    .get("content-type")
    ?.toLowerCase()
    ?.split(";")[0]
    ?.trim()

  if (
    !contentMediaType ||
    /application\/(?:.*\+)?json/.test(contentMediaType)
  ) {
    // application/json;
    // application/something+json;
    // application/vnd.something-other+json;

    // we opportunistically try to parse the body as json even if the content-type is not set
    // because it's not uncommon for APIs not to set any content-type and just send a json body
    try {
      const jsonBody = await clone.json()
      const contentSchema = schemaFromValue(jsonBody, "in:body")
      if (!contentSchema) return null
      return {
        type: "string" as const,
        contentSchema,
        contentMediaType: contentMediaType!,
      }
    } catch (e) {
      // Ignore this request.
      return null
    }
  } else if (/application\/(?:.*\+)?xml/.test(contentMediaType ?? "")) {
    // application/xml;
    // application/something+xml;
    // application/vnd.something-other+xml;
    // todo add xml parsing
    return { type: "string" as const, contentMediaType: contentMediaType! }
  } else if (contentMediaType?.includes("text/")) {
    return { type: "string" as const, contentMediaType: contentMediaType! }
  }
  // todo add other types

  // unknown type
  return null
}

const parseUrl = (request: Request) => {
  const urlObject = new URL(request.url, "http://localhost")

  if (request.headers.get("x-forwarded-host")) {
    urlObject.host = request.headers.get("x-forwarded-host")!
  }
  if (request.headers.get("x-forwarded-proto")) {
    urlObject.protocol = request.headers.get("x-forwarded-proto")!
  }
  if (request.headers.get("x-forwarded-port")) {
    urlObject.port = request.headers.get("x-forwarded-port")!
  }
  if (urlObject.hostname === "127.0.0.1") {
    urlObject.hostname = "localhost"
  }

  return urlObject
}

export const process = async ({
  request,
  response,
  direction,
}: {
  request: Request
  response: Response
  direction: "incoming" | "outgoing"
}): Promise<Operation> => {
  const requestBody = await getBodySchema(request)
  const responseBody = await getBodySchema(response)

  const requestHeadersSchemaEntries = [...request.headers.entries()]
    .map(([name, value]) => {
      const schema = schemaFromValue(value, "in:header") as
        | StringSchemaType
        | undefined
      return schema ? ([name, schema] as const) : undefined
    })
    .filter(isNonNullable)

  const responseHeadersSchemaEntries = [...response.headers.entries()]
    .map(([name, value]) => {
      const schema = schemaFromValue(value, "in:header") as
        | StringSchemaType
        | undefined
      return schema ? ([name, schema] as const) : undefined
    })
    .filter(isNonNullable)

  const urlObject = parseUrl(request)

  const query = [...urlObject.searchParams.entries()]
    .map(([name, value]) => {
      const schema = schemaFromValue(value, "in:query")
      return schema ? [name, schema] : undefined
    })
    .filter(isNonNullable)

  return {
    direction,
    request: {
      method: request.method,
      uri: urlObject.href.split("?")[0]!,
      headers: Object.fromEntries(requestHeadersSchemaEntries),
      query: Object.fromEntries(query),
      body: requestBody,
    },
    response: {
      headers: Object.fromEntries(responseHeadersSchemaEntries),
      statusCode: response.status,
      body: responseBody,
    },
  }
}
