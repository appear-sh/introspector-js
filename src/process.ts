import { ResolvedAppearConfig } from "./config"
import { schemaFromValue } from "./contentTypes/contentTypes"
import { StringSchemaType } from "./contentTypes/jsonSchema.types"
import { isNonNullable } from "./helpers"
import { Operation } from "./report"

const getBodySchema = async (input: Request | Response) => {
  const clone = input.clone()
  if (!clone.body) return null

  const contentMediaType = clone.headers
    .get("content-type")
    ?.toLowerCase()
    ?.split(";")[0]
    ?.trim()

  if (/application\/(?:.*\+)?json/.test(contentMediaType ?? "")) {
    // application/json;
    // application/something+json;
    // application/vnd.something-other+json;

    // We need to try/catch the `.json()` call because even though the
    // content-type is application/json the body might not be valid json.
    try {
      const jsonBody = await clone.json()
      const contentSchema = schemaFromValue(jsonBody, "in:body")
      if (!contentSchema) return null
      return {
        type: "string" as const,
        contentSchema,
        contentMediaType: contentMediaType!,
      }
    } catch (e) {
      // Ignore this request.
      // TODO: Eventually log this as an error.
      return null
    }
  } else if (/application\/(?:.*\+)?xml/.test(contentMediaType ?? "")) {
    // application/xml;
    // application/something+xml;
    // application/vnd.something-other+xml;
    // todo add xml parsing
    return { type: "string" as const, contentMediaType: contentMediaType! }
  } else if (contentMediaType?.includes("text/")) {
    return { type: "string" as const, contentMediaType: contentMediaType! }
  }
  // todo add other types

  // unknown type
  return null
}

export const process = async (
  req: Request,
  res: Response,
  config: ResolvedAppearConfig,
): Promise<Operation> => {
  const requestBody = await getBodySchema(req)
  const responseBody = await getBodySchema(res)

  const requestHeadersSchemaEntries = [...req.headers.entries()]
    .map(([name, value]) => {
      const schema = schemaFromValue(value, "in:header") as
        | StringSchemaType
        | undefined
      return schema ? ([name, schema] as const) : undefined
    })
    .filter(isNonNullable)

  const responseHeadersSchemaEntries = [...res.headers.entries()]
    .map(([name, value]) => {
      const schema = schemaFromValue(value, "in:header") as
        | StringSchemaType
        | undefined
      return schema ? ([name, schema] as const) : undefined
    })
    .filter(isNonNullable)

  const query = [...new URL(req.url, "http://localhost").searchParams.entries()]
    .map(([name, value]) => {
      const schema = schemaFromValue(value, "in:query")
      return schema ? [name, schema] : undefined
    })
    .filter(isNonNullable)

  return {
    request: {
      method: req.method,
      uri: req.url.split("?")[0]!, // remove query so we don't send raw values
      headers: Object.fromEntries(requestHeadersSchemaEntries),
      query: Object.fromEntries(query),
      body: requestBody,
    },
    response: {
      headers: Object.fromEntries(responseHeadersSchemaEntries),
      statusCode: res.status,
      body: responseBody,
    },
  }
}
