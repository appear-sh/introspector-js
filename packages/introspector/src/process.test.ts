import { describe, it, expect } from "vitest"
import { process } from "./process.js"

describe("process", () => {
  describe("request url normalization", () => {
    it("should normalize localhost", async () => {
      const req = new Request("http://127.0.0.1:3000/api/test?foo=bar")
      const op = await process({
        request: req,
        response: new Response(),
        direction: "incoming",
      })
      expect(op.request.uri).toBe("http://localhost:3000/api/test")
    })

    it("should remove unnecessary ports", async () => {
      const req = new Request("https://localhost:443/api/test?foo=bar")
      const op = await process({
        request: req,
        response: new Response(),
        direction: "incoming",
      })
      expect(op.request.uri).toBe("https://localhost/api/test")
    })

    it("should normalize forwarded headers", async () => {
      const req = new Request("http://127.0.0.1:3000/api/test?foo=bar", {
        headers: {
          "x-forwarded-host": "example.com",
          "x-forwarded-proto": "https",
          "x-forwarded-port": "443",
        },
      })
      const op = await process({
        request: req,
        response: new Response(),
        direction: "incoming",
      })
      expect(op.request.uri).toBe("https://example.com/api/test")
    })

    it("should use host header if it's set", async () => {
      const req = new Request("http://127.0.0.1:3000/api/test?foo=bar", {
        headers: { host: "example.com:4000" },
      })
      const op = await process({
        request: req,
        response: new Response(),
        direction: "incoming",
      })
      expect(op.request.uri).toBe("http://example.com:4000/api/test")
    })
  })

  describe("body schema", () => {
    it("should process a simple JSON request/response", async () => {
      const req = new Request("http://127.0.0.1:3000/api/test?foo=bar", {
        method: "POST",
        headers: { "content-type": "application/json", "x-custom": "abc" },
        body: JSON.stringify({ hello: "world" }),
      })
      const res = new Response(JSON.stringify({ ok: true }), {
        status: 200,
        headers: { "content-type": "application/json", "x-res": "xyz" },
      })

      const op = await process({
        request: req,
        response: res,
        direction: "incoming",
      })

      expect(op).toMatchSnapshot()
      expect(op.direction).toBe("incoming")
      expect(op.request.method).toBe("POST")
      expect(op.request.uri).toBe("http://localhost:3000/api/test")
      expect(op.request.headers).toHaveProperty("x-custom")
      expect(op.request.query).toHaveProperty("foo")
      expect(op.request.body).toMatchObject({
        type: "string",
        contentMediaType: "application/json",
      })
      expect(op.response.statusCode).toBe(200)
      expect(op.response.headers).toHaveProperty("x-res")
      expect(op.response.body).toMatchObject({
        type: "string",
        contentMediaType: "application/json",
      })
    })

    it("should handle missing content-type and parse as JSON if possible", async () => {
      const req = new Request("http://localhost/api/other", {
        method: "POST",
        body: JSON.stringify({ foo: 123 }),
      })
      const res = new Response("not json")
      const op = await process({
        request: req,
        response: res,
        direction: "outgoing",
      })
      expect(op.request.body).toMatchObject({
        type: "string",
        contentMediaType: "text/plain",
      })
      expect(op.response.body).toMatchObject({
        type: "string",
        contentMediaType: "text/plain",
      })
    })

    it("should handle text response", async () => {
      const req = new Request("http://localhost/api/text", { method: "GET" })
      const res = new Response("hello", {
        status: 200,
        headers: { "content-type": "text/plain" },
      })
      const op = await process({
        request: req,
        response: res,
        direction: "incoming",
      })
      expect(op.response.body).toMatchObject({
        type: "string",
        contentMediaType: "text/plain",
      })
    })
  })
})
