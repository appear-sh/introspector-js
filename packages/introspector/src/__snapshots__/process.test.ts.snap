// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`process > body schema > should process a simple JSON request/response 1`] = `
{
  "direction": "incoming",
  "request": {
    "body": {
      "contentMediaType": "application/json",
      "contentSchema": {
        "properties": {
          "hello": {
            "maxLength": 5,
            "minLength": 5,
            "type": "string",
          },
        },
        "required": [
          "hello",
        ],
        "type": "object",
      },
      "type": "string",
    },
    "headers": {
      "content-type": {
        "maxLength": 16,
        "minLength": 16,
        "type": "string",
      },
      "x-custom": {
        "maxLength": 3,
        "minLength": 3,
        "type": "string",
      },
    },
    "method": "POST",
    "query": {
      "foo": {
        "maxLength": 3,
        "minLength": 3,
        "type": "string",
      },
    },
    "uri": "http://localhost:3000/api/test",
  },
  "response": {
    "body": {
      "contentMediaType": "application/json",
      "contentSchema": {
        "properties": {
          "ok": {
            "type": "boolean",
          },
        },
        "required": [
          "ok",
        ],
        "type": "object",
      },
      "type": "string",
    },
    "headers": {
      "content-type": {
        "maxLength": 16,
        "minLength": 16,
        "type": "string",
      },
      "x-res": {
        "maxLength": 3,
        "minLength": 3,
        "type": "string",
      },
    },
    "statusCode": 200,
  },
}
`;
